shiny/ui/_theme.py

```diff
@@ -0,0 +1,513 @@
+    Create a custom Shiny theme.
+
+    The `Theme` class allows you to create a custom Shiny theme by providing custom Sass
+    code. The theme can be based on one of the available presets, such as `"shiny"` or
+    `"bootstrap"`, or a Bootswatch theme. Use the `.add_*()` methods can be chained
+    together to add custom Sass functions, defaults, mixins, and rules.
+
+    Pass the `Theme` object directly to the `theme` argument of any Shiny page function,
+    such as :func:`~shiny.ui.page_sidebar` or :func:`~shiny.ui.page_navbar`. In Shiny
+    Express apps, use the `theme` argument of :func:`~shiny.express.ui.page_opts` to set
+    the app theme.
+
+    **Note: Compiling custom themes requires the
+    [libsass](https://pypi.org/project/libsass/) package**, which is not installed by
+    default with Shiny. Use `pip install libsass` or `pip install shiny[theme]` to
+    install it.
+
+    Customized themes rare compiled to CSS when the theme is used. The `Theme` class
+    caches the compiled CSS so that it's only compiled for the first user to load your
+    app, but you can speed up app loading (and avoid the runtime `libsass` dependency)
+    by pre-compiling the theme CSS and saving it to a file. To do this, use the
+    `.to_css()` method to render the theme to a single minified CSS string. Once saved
+    to a file, the CSS can be used in any Shiny app by passing the file path to the
+    `theme` argument instead of the `Theme` object.
+
+    ```{.python filename="my_theme.py"}
+    from shiny import ui
+
+    my_theme = (
+        ui.Theme("shiny")
+        .add_mixins(
+            headings_color="$success",
+            bar_color="$purple",
+            select_color_text="$orange",
+        )
+    )
+
+    with open("my_theme.css", "w") as f:
+        f.write(my_theme.to_css())
+    ```
+
+    ```{.python filename="app.py"}
+    from shiny import ui
+
+    ui = ui.page_fluid(
+        # App content...
+        title="My App",
+        theme="my_theme.css",
+    )
+    ```
+
+    Parameters
+    ----------
+    preset
+        The name of the preset to use as a base. `"shiny"` is the default theme for
+        Shiny apps and `"bootstrap"` uses standard Bootstrap 5 styling. Bootswatch theme
+        presets are also available. Use `Theme.available_presets()` to see the full
+        list.
+    name
+        A custom name for the theme. If not provided, the preset name will be used.
+    include_paths
+        Additional paths to include when looking for Sass files used in `@import`
+        statements in the theme. This can be a single path as a string or
+        :class:`pathlib.Path`, or a list of paths. The paths should point to directories
+        containing additional Sass files that the theme depends on.
+
+    Raises
+    ------
+    ValueError
+        If the `preset` is not a valid theme preset.
+    """
+
+    def __init__(
+        self,
+        preset: ShinyThemePreset = "shiny",
+        name: Optional[str] = None,
```

Comment:
And similarly throughout the rest of the types

```suggestion
        name: str | None = None,
```

Reply:
I picked this stylistic choice up from other parts of our codebase where we use `Optional[str] = None` as a stronger signal that a function parameter is optional. I did a quick code search and found 648 uses of `Optional\[.+\] = None` in py-shiny.

Reply:
That's fine, but just want to add that, AFAIK, most/all of the `Optional` usage is there for historical reasons (we had to use it when supporting older versions of Python, but we could probably get rid of it now). Not necessarily because we're using it as an intentional signal of an optional argument.

Reply:
> Not necessarily because we're using it as an intentional signal of an optional argument.

I was trying to leave him out of it, but I had a pairing session with Barret a few months ago where he made this recommendation and I've been intentionally using `Optional[str] = None` as in this case since.

Reply:
Good to know, thanks!